Autopilot/Src/IMU.cpp

```diff
@@ -0,0 +1,234 @@
+   spi_port->set_slave(imuSlaveIdentifier);
+   spi_port->exchange_data(writeData, imuPlaceholderByte, 2);                                                                  
+}
+
+void ICM20602::get_accel_temp_gyro_reading() {
+   uint8_t imuBufferRead_FourteenBytes[15];
+   imuBufferRead_FourteenBytes[0] = REG_ACCEL_XOUT_H | 0x80;
+
+   uint8_t imu_raw_data[15];
+   imu_raw_data[0] = 0x00; //Dummy byte for MOSI
+   
+   //Store both High and Low Byte values
+   int16_t shiftedSensorAccX, shiftedSensorAccY, shiftedSensorAccZ, shiftedSensorTemp, shiftedSensorGyroX, shiftedSensorGyroY, shiftedSensorGyroZ; 
+   
+   for(int i = 1; i < 15; i++) {  
+      //0x00 is the reset value for all registers (Page 31). Thus, 0xFF will be used for the sensor_read (read only) registers
+      imuBufferRead_FourteenBytes[i] = 0xFF; //Initializes all dummy bytes in the MISO  
+      imu_raw_data[i] = 0x00; 
+   }
+   
+   //Uses the sensor registers to get raw data for all sensors
+   spi_port->set_slave(imuSlaveIdentifier);
+   sensorSuccess = spi_port->exchange_data(imuBufferRead_FourteenBytes, imu_raw_data, 15); //Sending in 15 bytes to tx_data (One register byte and fourteen dummy bytes)
+   
+   shiftedSensorAccX = (imu_raw_data[1] << 8) + imu_raw_data[2];
+   shiftedSensorAccY = (imu_raw_data[3] << 8) + imu_raw_data[4];
+   shiftedSensorAccZ = (imu_raw_data[5] << 8) + imu_raw_data[6];
+
+   shiftedSensorTemp = (imu_raw_data[7] << 8) + imu_raw_data[8];
+
+   shiftedSensorGyroX = (imu_raw_data[9] << 8) + imu_raw_data[10];
+   shiftedSensorGyroY = (imu_raw_data[11] << 8) + imu_raw_data[12];
+   shiftedSensorGyroZ = (imu_raw_data[13] << 8) + imu_raw_data[14];
+
+   //Converts 16-bit integers to a float. These are the actual measurements
+   measuredAccX = ((float) shiftedSensorAccX)/accelConversionFactor; 
+   measuredAccY = ((float) shiftedSensorAccY)/accelConversionFactor; 
+   measuredAccZ = ((float) shiftedSensorAccZ)/accelConversionFactor; 
+
+   measuredTemp = ((float) shiftedSensorTemp)/tempConversionFactor;
+
+   measuredGyroX = ((float) shiftedSensorGyroX)/gyroConversionFactor;
+   measuredGyroY = ((float) shiftedSensorGyroY)/gyroConversionFactor;
+   measuredGyroZ = ((float) shiftedSensorGyroZ)/gyroConversionFactor;
+}
+ 
+void ICM20602::Begin_Measuring() {
+   get_accel_temp_gyro_reading();
+   dataIsNew = true;
+}
+ 
+void ICM20602::GetResult(IMUData_t &Data) {
+   if (sensorSuccess == STATUS_CODE_OK) {
+      Data.isDataNew = dataIsNew;
+      dataIsNew = false;
+      Data.accx = measuredAccX;
+      Data.accy = measuredAccY;
+      Data.accz = measuredAccZ;
+      Data.gyrx = measuredGyroX;
+      Data.gyry = measuredGyroY;
+      Data.gyrz = measuredGyroZ;
+      Data.magx = NAN;
+      Data.magy = NAN;
+      Data.magz = NAN;
+      Data.temp = measuredTemp;
+      Data.sensorStatus = 0;
+   } else { //In case data was not received. Stick with old data
+      Data.isDataNew = false;
+      Data.sensorStatus = -1; 
+   }
```

Comment:
Note to future Dhruv when debugging the driver:

See if making sensorStatus -1 is needed. If the status is not 0, AttitudeManager will go into "fatal failure mode" which is NOT GOOD.